### PR TITLE
Add panic recovery within endLease

### PR DESCRIPTION
**Pull request by s-fairchild**
This is to ensure EndLease always gets called, including cases when a panic is encountered.

### Which issue this PR addresses:

[ARO-9241](https://issues.redhat.com/browse/ARO-9241)

Fixes

### What this PR does / why we need it:

In the event that a panic occurs within `endLease()`, we want to ensure that `ocb.dbOpenShiftClusters.EndLease()` is always called.

### Test plan for issue:

Run a local RP and induce a panic during a cluster operation.

### Is there any documentation that needs to be updated for this PR?

No.

### How do you know this will function as expected in production?

Local testing.

### CODE REVIEW

File: pkg/backend/openshiftcluster.go [hunk 1]
```diff
@@ -292,6 +292,12 @@ func (ocb *openShiftClusterBackend) endLease(ctx context.Context, log *logrus.En
 	var adminUpdateError *string
 	var failedProvisioningState api.ProvisioningState
 	initialProvisioningState := doc.OpenShiftCluster.Properties.ProvisioningState
+	var err error
+
+	defer func() {
+		recover.Panic(log)
+		_, err = ocb.dbOpenShiftClusters.EndLease(ctx, doc.Key, provisioningState, failedProvisioningState, adminUpdateError)
```
Comment (SudoBrendan) [inline pkg/backend/openshiftcluster.go:299]: I don't think this works as expected - `err` is swallowed in the `defer` here.

Here's a playground to show what I mean: https://go.dev/play/p/3Efb6cUNo8U

**Review by SudoBrendan — CHANGES_REQUESTED** [review]

File: pkg/backend/openshiftcluster.go (hunk 1 above)
Comment (SudoBrendan) [reply pkg/backend/openshiftcluster.go:299]: found the fix - use a named return result, in the playground, changing the function header to this works:

func sevenIsTooScary() (err error)

in the `defer`, we are then modifying the named `err` value.

tl;dr - remove `var err error` and change the `endLease` header to return `(err error)` instead of `error`.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]

**Review by bitoku — APPROVED** [review]

File: pkg/backend/openshiftcluster.go (hunk 1 above)
Comment (bitoku) [reply pkg/backend/openshiftcluster.go:298]: @kimorris27 
This might help
https://news.ycombinator.com/item?id=14668955
> Yep, and one cannot simulate that with local variables. In Go "return v" copies v into the return location before calling the deferred code. If that location is not named, the deferred function has no way to change it

Comment (kimorris27) [reply pkg/backend/openshiftcluster.go:298]: Yeah, that does help, thank you! I learned some new things about Go today.

**Review by kimorris27 — APPROVED** [review]

**Review by mociarain — APPROVED** [review]

**Review by rajdeepc2792 — APPROVED** [review]

**Review by rhamitarora — APPROVED** [review]

**Review by tsatam — APPROVED** [review]

**Review by SudoBrendan — APPROVED** [review]
